### PR TITLE
Cpu lstm inference

### PR DESCRIPTION
**Pull request by Jerryzcn**
## Description ##
(Brief description on what this PR is about)
CPU LSTM inference kernel. 
This is around 9.5x faster than gluon LSTM cell.

Verified on speech recognition task, as well as on unittest.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
@Jerryzcn It's very great to explore the full CPU power for RNN cell.

FYI, we have implemented the fused LSTM OP of CPU in local including inference and training.
And the new OP is registered with NNVM. 

I think we can co-operate together to merge the code.  @TaoLv @Sherry-Zhang
We can PR our code in your repo so that the whole LSTM solution can be ready in the same time.

What's your opinion?

**Comment by Jerryzcn** [comment]
@pengzhao-intel This is great, we can definitely collaborate.  The reason I am sending this PR is for one of our own project. and we would like to have something to use ASAP. Do you have a timeline for the fused LSTM OP?

**Comment by reminisce** [comment]
It's great to have cpu version implemented. We are deprecating operator implementation using the legacy interface. It's would be better if you can refactor code using the nnvm interface for operator implementation. One example is https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/convolution-inl.h#L155

**Comment by pengzhao-intel** [comment]
@Jerryzcn Thanks for the info. So, I think it's better to merge this PR as-is.

Our LSTM/GRU will be ready in this month and we will submit the code separately for the review :)

**Comment by szha** [comment]
@pengzhao-intel will the elman RNN be part of your PR? cudnn currently supports it, so we support it as part of the RNN layers.

**Comment by TaoLv** [comment]
@szha To make it easier to review, we would like to split the whole RNN implementation on CPU into several PRs. Firstly, We will submit code for single-layer and unidirectional LSTM/GRU. Then, multi-layer and bidirectional support will be added for LSTM/GRU. Vanilla RNN (maybe elman RNN in your words) will be supported after we finish LSTM/GRU. Actually, we have already done the code for fused vanilla RNN, but I think it should be a low priority compared with LSTM/GRU.

@szha What about your opinion? We can set a detailed plan for this PRs if needed.
@pengzhao-intel Correct me if I missed anything.

**Comment by szha** [comment]
@TaoLv sounds good. What timeline are we looking at for feature parity with cudnn?

**Comment by TaoLv** [comment]
@szha Team need take an internal discussion about it and will back to you with a detailed plan soon.
BTW, do you know anybody can help to refactor the existing RNN operator with nnvm interfaces? Seems that some of gpu code need be changed for that.

**Comment by szha** [comment]
Pinging @piiswrong for coordination.

**Comment by piiswrong** [comment]
We don't have to do the refactor now. CPU support is more important.

File: python/mxnet/gluon/data/dataloader.py [hunk 1]
```diff
@@ -214,6 +214,7 @@ def __iter__(self):
             worker.start()
             workers.append(worker)
 
+        idx = -1
```
Comment (szha) [inline python/mxnet/gluon/data/dataloader.py]: ?

Comment (Jerryzcn) [reply python/mxnet/gluon/data/dataloader.py]: idx might be reference before assignment when i run pylint

**Comment by CodingCat** [comment]
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the JIRA id

Thanks!

File: src/operator/rnn-inl.h [hunk 2]
```diff
@@ -78,10 +82,12 @@ inline int rnn_param_size(int layerNum,
...
   // get size of remaining layers
   if (bidirectional) {
-    size += (layerNum - 1) * rnn_single_param_size(2 * hiddenSize, hiddenSize, mode);
+    size += (layerNum - 1) * rnn_single_param_size(2 * hiddenSize,
+        hiddenSize, mode);
     size *= 2;
   } else {
-    size += (layerNum - 1) * rnn_single_param_size(hiddenSize, hiddenSize, mode);
+    size += (layerNum - 1) * rnn_single_param_size(hiddenSize, hiddenSize,
+        mode);
```
[the remark this answers is not in the document]
Comment (Jerryzcn) [reply src/operator/rnn-inl.h]: yes

File: src/operator/rnn-inl.h [hunk 3]
```diff
@@ -287,8 +507,9 @@ class RNNProp : public OperatorProperty {
     const std::vector<int> &out_grad,
     const std::vector<int> &in_data,
     const std::vector<int> &out_data) const override {
-    std::vector<int> dep = {in_data[rnn_enum::kData], in_data[rnn_enum::kParams],
-        in_data[rnn_enum::kState], out_data[rnn_enum::kOut], out_grad[rnn_enum::kOut]};
+    std::vector<int> dep = {in_data[rnn_enum::kData],
+  in_data[rnn_enum::kParams], in_data[rnn_enum::kState],
+  out_data[rnn_enum::kOut], out_grad[rnn_enum::kOut]};
```
[the remark this answers is not in the document]
Comment (Jerryzcn) [reply src/operator/rnn-inl.h]: it exceeded 80 char per line limit.

**Comment by Jerryzcn** [comment]
Network: 4 layer biLSTM with 320 hidden unit, with input shape of 800,1,40. 
Machine: C4.2xlarge OMP_NUM_THREADS=4 Ubuntu
uname -a
Linux ip-xxx 4.4.0-1052-aws #61-Ubuntu SMP Mon Feb 12 23:05:58 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
After warm up of 3 iteration, the average time over 10 iteration is 1.15sec

**Comment by szha** [comment]
Maybe provide a code snippet for the speed test?

**Comment by Jerryzcn** [comment]
`x = nd.ones(shape=(800,1,40))

model = mx.gluon.nn.Sequential()
with model.name_scope():
    model.add(mx.gluon.rnn.LSTM(320, num_layers=4, bidirectional=True))
model.initialize(mx.init.One())
for i in range(3):
    y = model(x).asnumpy()
    print(i)

t = time.time()
for i in range(10):
    y = model(x).asnumpy()
    print(i)
print("avg time:", (time.time() - t)/10.0)`

File: src/operator/rnn-inl.h [hunk 4]
```diff
@@ -114,7 +120,8 @@ struct RNNParam : public dmlc::Parameter<RNNParam> {
 
     DMLC_DECLARE_FIELD(p).set_default(0.)
     .set_range(0, 1)
-    .describe("Dropout probability, fraction of the input that gets dropped out at training time");
+    .describe("Dropout probability, fraction of the input that gets dropped"
+        "out at training time");
```
Comment (TaoLv) [inline src/operator/rnn-inl.h]: Remove this change. Length of this line is less than 100.
BTW, why there are still some parameters don't have their descriptions, like pkeep_, lstm_q_?

File: src/operator/rnn-inl.h [hunk 5]
```diff
@@ -144,15 +149,223 @@ class RNNOp : public Operator {
...
+          mshadow::Shape2(
+              y.shape_[0] * y.shape_[1], y.shape_[2]), s);  // (T*N)C
+
+    CHECK_EQ(x.CheckContiguous(), true);
+    CHECK_EQ(w.CheckContiguous(), true);
+    CHECK_EQ(hx.CheckContiguous(), true);
+    CHECK_EQ(y.CheckContiguous(), true);
+
+    CHECK(!ctx.is_train) << "only inference mode is available"
+      "for cpu at the moment.";
```
Comment (TaoLv) [inline src/operator/rnn-inl.h]: How about check this at the front of this function?

File: src/operator/rnn-inl.h (hunk 4 above)
Comment (Sherry-Zhang) [reply src/operator/rnn-inl.h]: pkeep_, lstm_q_ are used in cudnn_rnn-inl.h

File: tests/python/unittest/test_gluon_rnn.py [hunk 6]
```diff
@@ -67,6 +67,22 @@ def test_lstm_forget_bias():
...
     assert_allclose(mod.get_params()[0][bias_argument].asnumpy(), expected_bias)
 
+def test_lstm_cpu_inference():
+    # should behave the same as lstm cell
+    atol = 1e-6
+    x = mx.nd.ones(shape=(2, 2, 2))
+    model = mx.gluon.nn.Sequential()
+    with model.name_scope():
+        model.add(mx.gluon.rnn.LSTM(2, num_layers=6, bidirectional=True))
+    model.initialize(mx.init.One())
```
Comment (szha) [inline tests/python/unittest/test_gluon_rnn.py]: Could you also test the consistency between cpu and gpu, with same random weights and random inputs?

File: src/operator/rnn-inl.h [hunk 7]
```diff
@@ -144,15 +149,223 @@ class RNNOp : public Operator {
...
+          mshadow::Shape2(seq_len * batch_size, in_channel), s);  // (T*N)C
+    Tensor<cpu, 2, DType> y_flatten = out_data[rnn_enum::kOut]
+      .get_with_shape<cpu, 2, DType>(
+          mshadow::Shape2(
+              y.shape_[0] * y.shape_[1], y.shape_[2]), s);  // (T*N)C
+
+    CHECK_EQ(x.CheckContiguous(), true);
+    CHECK_EQ(w.CheckContiguous(), true);
+    CHECK_EQ(hx.CheckContiguous(), true);
+    CHECK_EQ(y.CheckContiguous(), true);
```
Comment (szha) [inline src/operator/rnn-inl.h]: CHECK(x.CheckContiguous());

File: tests/python/unittest/test_gluon_rnn.py (hunk 6 above)
Comment (Jerryzcn) [reply tests/python/unittest/test_gluon_rnn.py]: will it break CPU tests?

File: src/operator/rnn-inl.h [hunk 8]
```diff
@@ -153,6 +157,214 @@ class RNNOp : public Operator {
...
+                        const std::vector<OpReqType> &req,
+                        const std::vector<TBlob> &in_grad,
+                        const std::vector<TBlob> &aux_args) {
+    LOG(FATAL) << "LSTM backward is not available for cpu at the moment.";
+  }
+
+ private:
+  RNNParam param_;
+
+  virtual void LSTMFusedElementWiseCPUOps(const Tensor<cpu, 2, DType> &i2h_y,
```
Comment (reminisce) [inline src/operator/rnn-inl.h]: Why `virtual`?

File: src/operator/rnn-inl.h [hunk 9]
```diff
@@ -153,6 +157,214 @@ class RNNOp : public Operator {
...
+                                          const int copy_tmp2y) {
+    int64_t ji;
+    #pragma omp parallel for private(ji)
+    for (ji = 0; ji < batch_size * h_channel; ji++) {
+      int64_t j = ji / h_channel;  // batch dim
+      int64_t i = ji % h_channel;
+      int64_t f = i + h_channel;
+      int64_t c = i + h_channel * 2;
+      int64_t o = i + h_channel * 3;
+      h2h_y[j][i] += i2h_y[j][i];
```
Comment (reminisce) [inline src/operator/rnn-inl.h]: Too many overloaded operator `[]` calls and temporary Tensor objects generated here. At least you can cache `h2h_y[j]`, `i2h_y[j]`, etc. for each loop.

File: tests/python/unittest/test_gluon_rnn.py (hunk 6 above)
Comment (reminisce) [reply tests/python/unittest/test_gluon_rnn.py]: You can use this function to test consistency.
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/test_utils.py#L1207

File: tests/python/unittest/test_gluon_rnn.py [hunk 10]
```diff
@@ -67,6 +67,22 @@ def test_lstm_forget_bias():
...
+    y = model(x).asnumpy()
+
+    mx.test_utils.assert_almost_equal(y, np.array([[[0.72045636, 0.72045636, 0.95215213, 0.95215213],
```
Comment (reminisce) [inline tests/python/unittest/test_gluon_rnn.py]: Why are there hardcoded numbers?

Comment (Jerryzcn) [reply tests/python/unittest/test_gluon_rnn.py]: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/test_utils.py#L1207 is for symbol

File: src/operator/rnn-inl.h (hunk 9 above)
Comment (Jerryzcn) [reply src/operator/rnn-inl.h]: just tried. did not notice any difference in runtime. i think the tensor object probably does not generate new tensor object here.

File: src/operator/rnn-inl.h [hunk 11]
```diff
@@ -153,6 +157,214 @@ class RNNOp : public Operator {
...
+                                  const Tensor<cpu, 2, DType> &cy,
+                                  const int64_t batch_size,
+                                  const int64_t h_channel,
+                                  const int64_t t,
+                                  const int reverse_dir,
+                                  const int copy_tmp2y) {
+    int64_t ji;
+    #pragma omp parallel for private(ji)
+    for (ji = 0; ji < batch_size * h_channel; ji++) {
+      int64_t j = ji / h_channel;  // batch dim
```
Comment (Sherry-Zhang) [inline src/operator/rnn-inl.h]: Is it ok to write `batch_size * h_channel` in condition expression? It will calculate `ji` times.
And `++ji` is better than `ji++`.

Comment (Sherry-Zhang) [inline src/operator/rnn-inl.h]: Don't need to set `ji` private if define `ji`  in for loop. like this:
```
#pragma omp parallel for
for(int64_t ji = 0; ... ; ....)
```

File: src/operator/rnn-inl.h (hunk 9 above)
Comment (reminisce) [reply src/operator/rnn-inl.h]: When you use `data[i]`, where data is a 2D tensor, it returns a 1D temporary Tensor object for you, and then call the 1D tensor's operator[]. You would not be able to notice much runtime improvement after you make the change if the program didn't run for a long time only for this loop and the improvement could be dwarfed by other factors that are major bottlenecks. At least, it's not a good practice to write C++ code like this.

File: tests/python/unittest/test_gluon_rnn.py (hunk 6 above)
Comment (szha) [reply tests/python/unittest/test_gluon_rnn.py]: Under the current input and weight, your test would still pass even if the weights are iterated backwards. Unfortunately it's not in an acceptable state.

File: src/operator/rnn-inl.h (hunk 11 above)
Comment (Jerryzcn) [reply src/operator/rnn-inl.h]: u mean move it out of condition expression?

File: src/operator/rnn-inl.h (hunk 9 above)
Comment (Jerryzcn) [reply src/operator/rnn-inl.h]: okay. but it seems inside mshadow, all the ops are implemented using multiple `[]`
https://github.com/dmlc/mshadow/blob/master/mshadow/tensor_cpu-inl.h#L380
I will probably access the dptr_

Comment (reminisce) [reply src/operator/rnn-inl.h]: I don't have a strong opinion on this. If you could use `dptr_`, that's the best for performance because it saves function calls and temp tensor object creation, but it could introduce the issue of code readability and defeat the purpose of OO.

I think the rule of thumb here is try to avoid temp tensor creation and destruction while keep the code readable. So  it's okay to use `operator[]` for a 1D Tensor since it only return values and cache the temp tensor created by calling `operator[]` for a 2D tensor.

**Comment by Jerryzcn** [comment]
@szha I think https://github.com/apache/incubator-mxnet/blob/master/tests/python/gpu/test_operator_gpu.py#L1527 check for consistency? Although the inputs are ones.

**Comment by szha** [comment]
Since this change is only useful for inference, RNN layer still needs to remain a Block. Once the backward is in place, we will be able to change it to a HybridBlock.
